kernel/src/process.rs

```
@@ -340,96 +340,64 @@ impl<'a> Process<'a> {
-                    0,
-                    mpu::ExecutePermission::ExecutionPermitted,
-                    mpu::AccessPermission::ReadWrite);
+        match MPU::create_region(1, data_start, data_len,
+                        mpu::ExecutePermission::ExecutionPermitted,
+                        mpu::AccessPermission::ReadWrite) {
+            None =>
+                panic!("Infeasible MPU allocation. Base {:#x}, Length: {:#x}",
+                           data_start, data_len),
+            Some(region) => mpu.set_mpu(region)
+        }
 
         // Disallow access to grant region
-        let grant_size = unsafe {
+        let grant_len = unsafe {
             math::PowerOfTwo::ceiling(
                 self.memory.as_ptr().offset(self.memory.len() as isize) as u32 -
                     (self.kernel_memory_break as u32)
-            )
+            ).as_num::<u32>()
         };
         let grant_base = unsafe {
             self.memory
                 .as_ptr()
                 .offset(self.memory.len() as isize)
-                .offset(-(grant_size.as_num::<u32>() as isize))
+                .offset(-(grant_len as isize))
         };
-        mpu.set_mpu(2,
-                    grant_base as u32,
-                    grant_size,
-                    0,
-                    mpu::ExecutePermission::ExecutionNotPermitted,
-                    mpu::AccessPermission::PrivilegedOnly);
+
+        match MPU::create_region(2, grant_base as usize, grant_len as usize,
+                                 mpu::ExecutePermission::ExecutionNotPermitted,
+                                 mpu::AccessPermission::PrivilegedOnly) {
+            None =>
+                panic!("Infeasible MPU allocation. Base {:#x}, Length: {:#x}",
+                           grant_base as usize, grant_len),
+            Some(region) => mpu.set_mpu(region)
+        }
 
         // Setup IPC MPU regions
         for (i, region) in self.mpu_regions.iter().enumerate() {
```

Yes.. with the caveat that `None` should probably `panic!` here in the interest of failing quickly for the unexpected case (by unexpected, I mean that IPC doesn't work when it's expected to. I think that the request to create a 6th (or N+1th) IPC should've failed gracefully before reaching this code path, which motivates panic'ing here)